README.rst
======
pairme
======


Facilitate mob programming

* pull random pairs (each person appears 2 times, once as a driver and once as a navigator)
* reminds to change pair after some time interval (default 15 minutes)


Usage
-----


.. code-block:: shell

    $ git clone git@github.com:lenarother/pairme.git
    $ cd pairme
    $ python -m venv venv
    $ pip install -r requirements_dev.txt

    $ Configure mob session
    $ src/pairme/cli.py --mobteam John,Jane,Bob
    $ src/pairme/cli.py --mobtime 600

    $ Start mob session
    $ src/pairme/cli.py


TODO
----

* test properly
* proper config path handling
* move README.rst to md
* option to just draw pairs
* sound config
* make command line tool (like e.g. black)


Credits
-------

This package was created with Cookiecutter_ and the `audreyr/cookiecutter-pypackage`_ project template.

.. _Cookiecutter: https://github.com/audreyr/cookiecutter
.. _`audreyr/cookiecutter-pypackage`: https://github.com/audreyr/cookiecutter-pypackage


HISTORY.rst
=======
History
=======

0.0.1 (2022-11-10)
------------------

* First release on PyPI.


src/tests/__init__.py
"""Unit test package for pairme."""


src/tests/test_cli.py
#!/usr/bin/env python
from click.testing import CliRunner

from src.pairme import cli


def test_command_line_interface():
    """Test the CLI."""
    pass
    # runner = CliRunner()
    # result = runner.invoke(cli.main)
    # assert result.exit_code == 0
    # assert "pairme.cli.main" in result.output
    # help_result = runner.invoke(cli.main, ["--help"])
    # assert help_result.exit_code == 0
    # assert "--help  Show this message and exit." in help_result.output


src/tests/test_pair.py
#!/usr/bin/env python
from string import ascii_uppercase
import pytest

from src.pairme.pair import pull_pairs

TEAM_ODD = ["Jane", "John", "Jack"]
TEAM_EVEN = ["Anne", "Ben", "Chuck", "Dorothy"]

TEAM_LONG = list(ascii_uppercase)
TEAM_SHORT = ["Jane", "John"]


@pytest.mark.parametrize("team", [TEAM_ODD, TEAM_EVEN, TEAM_LONG, TEAM_SHORT])
def test_right_number_of_pairs_is_pulled(team):
    pairs = pull_pairs(team)
    assert len(team) == len(pairs)


@pytest.mark.parametrize("team", [TEAM_ODD, TEAM_EVEN, TEAM_LONG, TEAM_SHORT])
def test_each_pair_contains_two_different_team_members(team):
    pairs = pull_pairs(team)
    assert all(len(set(pair)) == 2 for pair in pairs)


@pytest.mark.parametrize("team", [TEAM_ODD, TEAM_EVEN, TEAM_LONG, TEAM_SHORT])
def test_each_team_member_is_pulled_twice(team):
    pairs = pull_pairs(team)
    scrum_master = sorted([pair[0] for pair in pairs])
    backup_scrum_master = sorted([pair[1] for pair in pairs])
    assert sorted(team) == scrum_master == backup_scrum_master


def test_pairs_are_random():
    pairs1 = pull_pairs(TEAM_LONG)
    pairs2 = pull_pairs(TEAM_LONG)
    pairs3 = pull_pairs(TEAM_LONG)
    assert not (pairs1 == pairs2 == pairs3)


@pytest.mark.parametrize("team", [[], ["John"], ["John", "John"]])
def test_invalid_team_raises(team):
    with pytest.raises(ValueError):
        pull_pairs(team)


src/pairme/cli.py
#!/usr/bin/env python

"""Console script for pairme."""

import sys
import click

from src.pairme.pair import pull_pairs
from src.pairme.session import run_mob
from src.pairme import config


@click.command()
@click.option("--mobteam", type=str, help="Team members e.g. Bob,Jane,Jon")
@click.option("--mobtime", type=int, help="Time for one mob pair.")
def main(mobteam=None, mobtime=None):
    if mobteam:
        config.set_team(mobteam)
        click.echo(f"Team set to {mobteam}.")
        return

    if mobtime:
        config.set_time(mobtime)
        click.echo(f"Mob time round set to {mobtime}.")
        return

    team = config.get_team()
    interval = config.get_time()
    break_duration = config.get_break_duration()
    break_frequency = config.get_break_frequency()

    mob_pairs = pull_pairs(team)
    for pair in mob_pairs:
        click.echo(f"Driver: {pair[0]} - Navigator: {pair[1]}.")
    click.echo(f"-" * 50)
    run_mob(
        pairs=mob_pairs,
        time_interval=interval,
        break_duration=break_duration,
        break_frequency=break_frequency,
    )
    return


if __name__ == "__main__":
    sys.exit(main())  # pragma: no cover


src/pairme/session.py
import datetime
import os
import platform
import time

import beepy as beep
import click

PAIRME_SEPARATOR = "-" * 50
MINUTE = 60


def make_sound(*args):
    system = platform.platform().lower()
    if system.startswith("macos"):
        for message in args:
            os.system(f"say {message}")
    else:
        beep.beep(7)


def should_take_break(turns_counter, break_frequency):
    return turns_counter != 0 and turns_counter % break_frequency == 0


def take_break(break_duration):
    break_end = datetime.datetime.now() + datetime.timedelta(seconds=break_duration)
    break_message = (
        f"Take a {int(break_duration / MINUTE)} minutes break until {break_end:%H:%M}."
    )
    click.echo(break_message)
    click.echo(PAIRME_SEPARATOR)
    make_sound(break_message)

    time.sleep(break_duration)


def run_mob(pairs, time_interval, break_duration, break_frequency):
    pairs.reverse()
    for counter, pair in enumerate(pairs):
        if should_take_break(counter, break_frequency):
            take_break(break_duration)

        message_change = "Times up, change pair!"
        message_pair = f"Driver is {pair[0]}, navigator is {pair[1]}."

        click.echo(message_change)
        click.echo(message_pair)
        click.echo(PAIRME_SEPARATOR)

        make_sound(message_change, message_pair)

        time.sleep(time_interval)


src/pairme/config.py
from pathlib import Path
import configparser

PROJECT_DIR = Path(__file__).parent
CONFIG_FILE = f"{PROJECT_DIR}/.pairme.conf"
CONFIG_SECTION = "pairme"


def get_config():
    config = configparser.ConfigParser()
    config.read(CONFIG_FILE)
    return config


def write_config(config):
    with open(CONFIG_FILE, "w") as configfile:
        config.write(configfile)


def set_config_value(name, value):
    config = get_config()
    if not config.has_section(CONFIG_SECTION):
        config.add_section(CONFIG_SECTION)
    config.set(CONFIG_SECTION, name, value)
    write_config(config)


def get_config_value(name, fallback=None):
    config = get_config()
    if fallback:
        return config.get(CONFIG_SECTION, name, fallback=fallback)
    return config.get(CONFIG_SECTION, name)


def set_team(members_list):
    set_config_value("team", members_list)


def set_time(seconds):
    set_config_value("time", str(seconds))


def get_team():
    try:
        team = get_config_value("team")
        return [i.strip() for i in team.split(",")]
    except (configparser.NoSectionError, configparser.NoOptionError):
        raise KeyError("No team configured, run: cli.py --mobteam Jon,Jane,Ben...")


def get_time():
    return int(get_config_value("time", fallback=15 * 60))


def get_break_frequency():
    return int(get_config_value("break_frequency", fallback=2))


def get_break_duration():
    return int(get_config_value("break_duration", fallback=5 * 60))
